{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# __Step 4.1: Topic model__\n",
    "\n",
    "The kmean cluserting results are not particularly clear what's going on. So go stiraght to topic modeling.\n",
    "\n",
    "Goals:\n",
    "- Get topic models\n",
    "- Get topics\n",
    "\n",
    "The modeling run is done on HPC through `script_4_1_topic_model_v2.py`.\n",
    "- RuntimeError: CUDA error: no kernel image is available for execution on the device\n",
    "  - Fixed by creating a new bertopic environment on HPC with new install of pytorch.\n",
    "- /var/lib/slurmd/job58225495/slurm_script: line 19: 28351 Killed\n",
    "  - Unsure why. Error did not occur again.\n",
    "- max_topic = unique_topics[-1], IndexError: list index out of range\n",
    "  - Changed BERTopic setting without specifying min_topic_size or nr_topics. \n",
    "\n",
    "Topic model comparison\n",
    "- All three (distillbert, scibert, biobert) have similar top topics with similar wordings. Thus, even general purpose BERT after retraining with the corpus, it is possible to pick out the topics.\n",
    "- Among them, scibert has the least number of docs in the -1 (outlier) topic, so it is chosen as the embedding model for further analysis (4.2).\n",
    "- There is 268848 docs in the -1 topic. Use [the tips from BERTopic FAQ](https://maartengr.github.io/BERTopic/faq.html#how-do-i-reduce-topic-outliers) to fix it."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ___Set up___"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Module import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re, pickle, os\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "from nltk.corpus import stopwords\n",
    "from nltk.tokenize import word_tokenize\n",
    "from tqdm import tqdm\n",
    "from bertopic import BERTopic"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Key variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reproducibility\n",
    "seed = 20220609\n",
    "\n",
    "# Setting working directory\n",
    "proj_dir   = Path.home() / \"projects/plant_sci_hist\"\n",
    "work_dir   = proj_dir / \"4_topic_model/4_1_compare_models\"\n",
    "work_dir.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "os.chdir(work_dir)\n",
    "\n",
    "# plant science corpus\n",
    "dir25       = proj_dir / \"2_text_classify/2_5_predict_pubmed\"\n",
    "corpus_file = dir25 / \"corpus_plant_421658.tsv.gz\"\n",
    "\n",
    "# Output files\n",
    "\n",
    "## The list obj containing cleaned docs\n",
    "docs_clean_file  = work_dir / f\"docs_clean.pickle\"\n",
    "\n",
    "## model names \n",
    "model1 = \"distilbert-base-uncased\"\n",
    "model2 = \"allenai/scibert_scivocab_uncased\" # scibert\n",
    "model3 = \"dmis-lab/biobert-base-cased-v1.2\" # biobert\n",
    "\n",
    "model2_mod = \"-\".join(model2.split(\"/\"))\n",
    "model3_mod = \"-\".join(model3.split(\"/\"))\n",
    "\n",
    "## topic models\n",
    "model1_file = work_dir / f'topic_model_{model1}'\n",
    "model2_file = work_dir / f'topic_model_{model2_mod}'\n",
    "model3_file = work_dir / f'topic_model_{model3_mod}'\n",
    "\n",
    "## topics\n",
    "topic1_file = work_dir / f'topics_{model1}.pickle'\n",
    "topic2_file = work_dir / f'topics_{model2_mod}.pickle'\n",
    "topic3_file = work_dir / f'topics_{model3_mod}.pickle'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Proprecess corpus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "corpus_df = pd.read_csv(corpus_file, sep='\\t', compression='gzip')\n",
    "corpus_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_text(x):\n",
    "    x = str(x)\n",
    "    x = x.lower()\n",
    "    # Replace any non-alphanumric characters of any length\n",
    "    # Q: Not sure what the # character do.\n",
    "    x = re.sub(r'#[A-Za-z0-9]*', ' ', x)\n",
    "    # tokenize and rid of any token matching stop words\n",
    "    tokens = word_tokenize(x)\n",
    "    x = ' '.join([w for w in tokens if not w in stop_words_dict])\n",
    "    return x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "docs       = corpus_df['txt']\n",
    "stop_words = stopwords.words('english')\n",
    "stop_words_dict = {}\n",
    "for i in stop_words:\n",
    "  stop_words_dict[i] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "docs_clean = []\n",
    "for doc_idx in tqdm(range(len(docs))):\n",
    "  doc = docs[doc_idx]\n",
    "  docs_clean.append(clean_text(doc))\n",
    "len(docs_clean)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(docs_clean_file, \"wb\") as f:\n",
    "  pickle.dump(docs_clean, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ___Run BERTopic___"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe6c03c3-1b09-4a28-9131-e4f5f6cb8ad9",
   "metadata": {},
   "source": [
    "### Initialize\n",
    "\n",
    "- language: str = 'english'\n",
    "- top_n_words: int = 10\n",
    "  - The number of words per topic to extract. __Setting this too high can negatively impact topic embeddings__ as topics are typically best represented by at most __10 words__.\n",
    "- n_gram_range: Tuple[int, int] = (1, 1)\n",
    "  - The n-gram range for the CountVectorizer, between 1 and 3, otherwise memory issue.\n",
    "- min_topic_size: int = 10\n",
    "  - The minimum size of the topic.\n",
    "- nr_topics: Union[int, str] = None\n",
    "  - Specifying the number of topics will reduce the initial number of topics to the value specified.\n",
    "  - Use __\"auto\"__ to automatically reduce topics using HDBSCAN\n",
    "- calculate_probabilities: bool = False\n",
    "  - Whether to calculate the probabilities of all topics per document instead of the probability of the assigned topic per document.\n",
    "  - Will significantly increase computing time if True.\n",
    "- diversity: float = None\n",
    "  - Whether to use MMR to diversify the resulting topic representations.\n",
    "  - Value between 0 (no divresity) and 1 (very diverse).\n",
    "    - __Q: What does diversity mean here?__\n",
    "- seed_topic_list: List[List[str]] = None\n",
    "  - A list of seed words per topic to converge around.\n",
    "- embedding_model=None\n",
    "  - SentenceTransformers, Flair, Spacy, Gensim, USE (TF-Hub), or [these](https://www.sbert.net/docs/pretrained_models.html).\n",
    "  - Try to use `allenai-specter`.\n",
    "- umap_model: umap.umap_.UMAP = None\n",
    "- hdbscan_model: hdbscan.hdbscan_.HDBSCAN = None\n",
    "- vectorizer_model: sklearn.feature_extraction.text.CountVectorizer = None\n",
    "- verbose: bool = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Because of issue with the line:\n",
    "# max_topic = unique_topics[-1]\n",
    "# IndexError: list index out of range\n",
    "# The following is tweaked to rid of min_topic_size and nr_topics setting and\n",
    "# the run completed.\n",
    "#topic_model = BERTopic(calculate_probabilities=False,\n",
    "#                       n_gram_range=(1,2),\n",
    "#                       min_topic_size=1000, \n",
    "#                       nr_topics='auto',\n",
    "#                       embedding_model=model1,\n",
    "#                       verbose=True)\n",
    "topic_model1 = BERTopic(calculate_probabilities=False,\n",
    "                        n_gram_range=(1,2),\n",
    "                        embedding_model=model1,\n",
    "                        verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6c193f3-a948-460c-924d-7c1198643908",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Fit_transform\n",
    "\n",
    "Long run time. Switch to HPC."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "topics1 = topic_model1.fit_transform(docs_clean)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "  topic_model1.save(model1_file)\n",
    "  with open(topic1_file, \"wb\") as f:\n",
    "    pickle.dump(topics1, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ___Analysis of fitted model___"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### distllbert"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(pandas.core.frame.DataFrame, (1443, 3))"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topic_model1_loaded = BERTopic.load(model1_file)\n",
    "topic_info1 = topic_model1_loaded.get_topic_info()\n",
    "type(topic_info1), topic_info1.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Topic</th>\n",
       "      <th>Count</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-1</td>\n",
       "      <td>323846</td>\n",
       "      <td>-1_plant_plants_activity_growth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>4799</td>\n",
       "      <td>0_spectroscopic_nmr_elucidated_compounds</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2567</td>\n",
       "      <td>1_cd_pb_zn_soil</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>2024</td>\n",
       "      <td>2_qtls_qtl_traits_trait</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>2011</td>\n",
       "      <td>3_μm_conidia_first report_et al</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4</td>\n",
       "      <td>1960</td>\n",
       "      <td>4_resistance_rust_qtl_rust resistance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>5</td>\n",
       "      <td>1859</td>\n",
       "      <td>5_medium_regeneration_callus_ms medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>6</td>\n",
       "      <td>1456</td>\n",
       "      <td>6_drought_tolerance_stress_aba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>7</td>\n",
       "      <td>1455</td>\n",
       "      <td>7_transformation_agrobacterium_tumefaciens_med...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>8</td>\n",
       "      <td>1424</td>\n",
       "      <td>8_genomics_technologies_breeding_biology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>9</td>\n",
       "      <td>1226</td>\n",
       "      <td>9_virus_nucleotide sequence_orf_nucleotides</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>10</td>\n",
       "      <td>1223</td>\n",
       "      <td>10_glutathione_ascorbate_gsh_antioxidant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>11</td>\n",
       "      <td>1182</td>\n",
       "      <td>11_nectar_pollination_flowers_pollinator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>12</td>\n",
       "      <td>1138</td>\n",
       "      <td>12_pest_larvae_adults_trap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>13</td>\n",
       "      <td>1129</td>\n",
       "      <td>13_floral_flowering_meristem_mads</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>14</td>\n",
       "      <td>1118</td>\n",
       "      <td>14_reflectance_near infrared_spectral_infrared</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>15</td>\n",
       "      <td>1023</td>\n",
       "      <td>15_database_http_web_annotation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>16</td>\n",
       "      <td>957</td>\n",
       "      <td>16_glycans_xyloglucan_oligosaccharides_beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>17</td>\n",
       "      <td>896</td>\n",
       "      <td>17_rhizosphere_communities_microbial_bacterial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>18</td>\n",
       "      <td>879</td>\n",
       "      <td>18_crispr_cas9_editing_crispr cas9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Topic   Count                                               Name\n",
       "0      -1  323846                    -1_plant_plants_activity_growth\n",
       "1       0    4799           0_spectroscopic_nmr_elucidated_compounds\n",
       "2       1    2567                                    1_cd_pb_zn_soil\n",
       "3       2    2024                            2_qtls_qtl_traits_trait\n",
       "4       3    2011                    3_μm_conidia_first report_et al\n",
       "5       4    1960              4_resistance_rust_qtl_rust resistance\n",
       "6       5    1859             5_medium_regeneration_callus_ms medium\n",
       "7       6    1456                     6_drought_tolerance_stress_aba\n",
       "8       7    1455  7_transformation_agrobacterium_tumefaciens_med...\n",
       "9       8    1424           8_genomics_technologies_breeding_biology\n",
       "10      9    1226        9_virus_nucleotide sequence_orf_nucleotides\n",
       "11     10    1223           10_glutathione_ascorbate_gsh_antioxidant\n",
       "12     11    1182           11_nectar_pollination_flowers_pollinator\n",
       "13     12    1138                         12_pest_larvae_adults_trap\n",
       "14     13    1129                  13_floral_flowering_meristem_mads\n",
       "15     14    1118     14_reflectance_near infrared_spectral_infrared\n",
       "16     15    1023                    15_database_http_web_annotation\n",
       "17     16     957        16_glycans_xyloglucan_oligosaccharides_beta\n",
       "18     17     896     17_rhizosphere_communities_microbial_bacterial\n",
       "19     18     879                 18_crispr_cas9_editing_crispr cas9"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topic_info1.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "topic_info1.to_csv(str(model1_file) + \"_topics.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### scibert"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1697, 3)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topic_model2_loaded = BERTopic.load(model2_file)\n",
    "topic_info2 = topic_model2_loaded.get_topic_info()\n",
    "topic_info2.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Topic</th>\n",
       "      <th>Count</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-1</td>\n",
       "      <td>268848</td>\n",
       "      <td>-1_plant_species_plants_growth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>6592</td>\n",
       "      <td>0_spectroscopic_nmr_elucidated_compounds</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>6073</td>\n",
       "      <td>1_transcriptome_degs_unigenes_differentially e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>4369</td>\n",
       "      <td>2_cd_pb_cu_zn</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>2618</td>\n",
       "      <td>3_qtls_qtl_traits_quantitative trait</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4</td>\n",
       "      <td>2522</td>\n",
       "      <td>4_et al_first report_symptoms_et</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>5</td>\n",
       "      <td>2292</td>\n",
       "      <td>5_regeneration_medium_callus_ms medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>6</td>\n",
       "      <td>2168</td>\n",
       "      <td>6_virus_viruses_nucleotide sequence_coat protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>7</td>\n",
       "      <td>2056</td>\n",
       "      <td>7_tolerance_drought_salt_stress</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>8</td>\n",
       "      <td>1859</td>\n",
       "      <td>8_gene family_family_genome wide_wide identifi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>9</td>\n",
       "      <td>1674</td>\n",
       "      <td>9_promoter_gus_promoters_reporter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>10</td>\n",
       "      <td>1629</td>\n",
       "      <td>10_rust_resistance_qtl_rust resistance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>11</td>\n",
       "      <td>1607</td>\n",
       "      <td>11_actin_microtubule_microtubules_spindle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>12</td>\n",
       "      <td>1548</td>\n",
       "      <td>12_trypsin_inhibitor_purified_enzyme</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>13</td>\n",
       "      <td>1535</td>\n",
       "      <td>13_antioxidant_antioxidant activity_phenolic_e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>14</td>\n",
       "      <td>1468</td>\n",
       "      <td>14_virulence_conidiation_appressorium_pathogen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>15</td>\n",
       "      <td>1441</td>\n",
       "      <td>15_na_ca2_atpase_ca</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>16</td>\n",
       "      <td>1361</td>\n",
       "      <td>16_mirnas_mirna_micrornas_lncrnas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>17</td>\n",
       "      <td>1327</td>\n",
       "      <td>17_pest_hemiptera_traps_trap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>18</td>\n",
       "      <td>1311</td>\n",
       "      <td>18_transformation_agrobacterium_tumefaciens_ag...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Topic   Count                                               Name\n",
       "0      -1  268848                     -1_plant_species_plants_growth\n",
       "1       0    6592           0_spectroscopic_nmr_elucidated_compounds\n",
       "2       1    6073  1_transcriptome_degs_unigenes_differentially e...\n",
       "3       2    4369                                      2_cd_pb_cu_zn\n",
       "4       3    2618               3_qtls_qtl_traits_quantitative trait\n",
       "5       4    2522                   4_et al_first report_symptoms_et\n",
       "6       5    2292             5_regeneration_medium_callus_ms medium\n",
       "7       6    2168   6_virus_viruses_nucleotide sequence_coat protein\n",
       "8       7    2056                    7_tolerance_drought_salt_stress\n",
       "9       8    1859  8_gene family_family_genome wide_wide identifi...\n",
       "10      9    1674                  9_promoter_gus_promoters_reporter\n",
       "11     10    1629             10_rust_resistance_qtl_rust resistance\n",
       "12     11    1607          11_actin_microtubule_microtubules_spindle\n",
       "13     12    1548               12_trypsin_inhibitor_purified_enzyme\n",
       "14     13    1535  13_antioxidant_antioxidant activity_phenolic_e...\n",
       "15     14    1468  14_virulence_conidiation_appressorium_pathogen...\n",
       "16     15    1441                                15_na_ca2_atpase_ca\n",
       "17     16    1361                  16_mirnas_mirna_micrornas_lncrnas\n",
       "18     17    1327                       17_pest_hemiptera_traps_trap\n",
       "19     18    1311  18_transformation_agrobacterium_tumefaciens_ag..."
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topic_info2.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "topic_info2.to_csv(str(model2_file) + \"_topics.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### biobert"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1452, 3)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topic_model3_loaded = BERTopic.load(model3_file)\n",
    "topic_info3 = topic_model3_loaded.get_topic_info()\n",
    "topic_info3.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Topic</th>\n",
       "      <th>Count</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-1</td>\n",
       "      <td>293790</td>\n",
       "      <td>-1_plant_growth_activity_plants</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>3403</td>\n",
       "      <td>0_qtl_qtls_mapping_rust</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>2891</td>\n",
       "      <td>1_communities_community_rhizosphere_microbial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>2671</td>\n",
       "      <td>2_degs_unigenes_transcriptome_differentially e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>2669</td>\n",
       "      <td>3_transformation_agrobacterium_tumefaciens_med...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4</td>\n",
       "      <td>2510</td>\n",
       "      <td>4_aba_tolerance_overexpression_stress</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>5</td>\n",
       "      <td>2358</td>\n",
       "      <td>5_et al_first report_symptoms_conidia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>6</td>\n",
       "      <td>2265</td>\n",
       "      <td>6_gene family_family_genome wide_genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>7</td>\n",
       "      <td>2174</td>\n",
       "      <td>7_virus_viruses_nucleotide sequence_mosaic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>8</td>\n",
       "      <td>2145</td>\n",
       "      <td>8_cd_pb_zn_contaminated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>9</td>\n",
       "      <td>2106</td>\n",
       "      <td>9_medium_regeneration_callus_ms medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>10</td>\n",
       "      <td>1894</td>\n",
       "      <td>10_spectroscopic_elucidated_structures_nmr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>11</td>\n",
       "      <td>1847</td>\n",
       "      <td>11_cdna_sequence_amino_clone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>12</td>\n",
       "      <td>1822</td>\n",
       "      <td>12_virus_viral_rna_movement</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>13</td>\n",
       "      <td>1625</td>\n",
       "      <td>13_actin_microtubule_microtubules_cytoskeleton</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>14</td>\n",
       "      <td>1618</td>\n",
       "      <td>14_pretreatment_lignin_hydrolysis_pretreated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>15</td>\n",
       "      <td>1567</td>\n",
       "      <td>15_pollination_pollinator_flowers_pollinators</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>16</td>\n",
       "      <td>1440</td>\n",
       "      <td>16_mirnas_mirna_micrornas_lncrnas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>17</td>\n",
       "      <td>1296</td>\n",
       "      <td>17_genetic diversity_genetic_diversity_populat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>18</td>\n",
       "      <td>1157</td>\n",
       "      <td>18_glutathione_ascorbate_antioxidant_peroxidase</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Topic   Count                                               Name\n",
       "0      -1  293790                    -1_plant_growth_activity_plants\n",
       "1       0    3403                            0_qtl_qtls_mapping_rust\n",
       "2       1    2891      1_communities_community_rhizosphere_microbial\n",
       "3       2    2671  2_degs_unigenes_transcriptome_differentially e...\n",
       "4       3    2669  3_transformation_agrobacterium_tumefaciens_med...\n",
       "5       4    2510              4_aba_tolerance_overexpression_stress\n",
       "6       5    2358              5_et al_first report_symptoms_conidia\n",
       "7       6    2265             6_gene family_family_genome wide_genes\n",
       "8       7    2174         7_virus_viruses_nucleotide sequence_mosaic\n",
       "9       8    2145                            8_cd_pb_zn_contaminated\n",
       "10      9    2106             9_medium_regeneration_callus_ms medium\n",
       "11     10    1894         10_spectroscopic_elucidated_structures_nmr\n",
       "12     11    1847                       11_cdna_sequence_amino_clone\n",
       "13     12    1822                        12_virus_viral_rna_movement\n",
       "14     13    1625     13_actin_microtubule_microtubules_cytoskeleton\n",
       "15     14    1618       14_pretreatment_lignin_hydrolysis_pretreated\n",
       "16     15    1567      15_pollination_pollinator_flowers_pollinators\n",
       "17     16    1440                  16_mirnas_mirna_micrornas_lncrnas\n",
       "18     17    1296  17_genetic diversity_genetic_diversity_populat...\n",
       "19     18    1157    18_glutathione_ascorbate_antioxidant_peroxidase"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topic_info3.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "topic_info3.to_csv(str(model3_file) + \"_topics.tsv\", sep='\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.4 ('torch_default': conda)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4 (main, Mar 31 2022, 08:41:55) [GCC 7.5.0]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "4255f477c905e3cafd6d08b9a6d118445dbfbaff982fd1d9831280a79a13df35"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
